### PR TITLE
Go memory usage optimisations

### PR DESCRIPTION
**Pull request by pandurangpatil**
Restructured the code to isolate the common handling for main source code pre-processing to build the cache, downloaded dependencies and AST creation.

With this change, we now have created separate traits which handle, main source code pre-processing, processing of downloaded dependencies for caching and AST Creation along with their respective passes.

With this, we get the following benefits.

1. Clarity in the code as the processing is now isolated in different passes and traits.
2. Earlier we held the AstCreator instances for each file at the beginning of CPG generation while pre-processing the AST for building the metadata cache which can be used and referred to by the AST creator to resolve the types. Resuing that instance at the last stage inside AstCreationPass. Which blocks large amounts of memory which is noticeable with a big code base. With this change, we are creating an AstCreator instance and destroying it once the pass is done with its work.

### CODE REVIEW

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/CommonCacheBuilder.scala [hunk 1]
```diff
@@ -3,71 +3,20 @@ package io.joern.gosrc2cpg.astcreation
...
-      .name(fullyQualifiedPackage)
-      .fullName(fullyQualifiedPackage)
-      .filename(packageFolderPath)
-    val fakePackageTypeDecl =
-      typeDeclNode(rootNode, fullyQualifiedPackage, fullyQualifiedPackage, packageFolderPath, fullyQualifiedPackage)
-    Ast(namespaceBlock).withChild(Ast(fakePackageTypeDecl))
-  }
-
-  private def processPackageLevelGolbalVaraiblesAndConstants(json: Value): Unit = {
+  protected def processPackageLevelGolbalVaraiblesAndConstants(json: Value): Unit = {
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/CommonCacheBuilder.scala:19]:
```suggestion
  protected def processPackageLevelGlobalVariablesAndConstants(json: Value): Unit = {
```

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/DepdencySrcProcessor.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package io.joern.gosrc2cpg.astcreation
+
+import io.joern.gosrc2cpg.parser.ParserKeys
+import io.joern.x2cpg.ValidationMode
+import ujson.{Arr, Obj, Value}
+
+trait DepdencySrcProcessor(implicit withSchemaValidation: ValidationMode) { this: AstCreator =>
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/DepdencySrcProcessor.scala:7]:
```suggestion
trait DependencySrcProcessor(implicit withSchemaValidation: ValidationMode) { this: AstCreator =>
```

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/PackageCtorCreationPass.scala [hunk 3]
```diff
@@ -11,12 +11,16 @@ import scala.jdk.CollectionConverters.*
 
 class PackageCtorCreationPass(cpg: Cpg, config: Config, goGlobal: GoGlobal)
     extends ConcurrentWriterCpgPass[(String, Set[(Ast, String)])](cpg) {
-  override def generateParts(): Array[(String, Set[(Ast, String)])] =
-    goGlobal.pkgLevelVarAndConstantAstMap
+  override def generateParts(): Array[(String, Set[(Ast, String)])] = {
+    val parts = goGlobal.pkgLevelVarAndConstantAstMap
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/PackageCtorCreationPass.scala:15]: Could also use a `try...finally` to ensure this clears and not require a variable assignment.

For clarity, might help to make a case class for the `(String, Set[(Ast, String)])` types

Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/PackageCtorCreationPass.scala:15]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/CommonCacheBuilder.scala (hunk 1 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/CommonCacheBuilder.scala]: done

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/BasePassForAstProcessing.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import org.slf4j.{Logger, LoggerFactory}
+
+abstract class BasePassForAstProcessing(
+  cpg: Cpg,
+  astFiles: List[String],
+  config: Config,
+  goMod: GoModHelper,
+  goGlobal: GoGlobal,
+  tmpDir: File
+) extends ConcurrentWriterCpgPass[String](cpg) {
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/BasePassForAstProcessing.scala:20]: Note this PR: https://github.com/joernio/joern/pull/4689
```suggestion
) extends ForkJoinParallelCpgPass[String](cpg) {
```

**Review by DavidBakerEffendi — CHANGES_REQUESTED** [review]

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/BasePassForAstProcessing.scala (hunk 4 above)
Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/BasePassForAstProcessing.scala:20]: @DavidBakerEffendi is there any way I can load test the new flatgraph db version along with 
`ForkJoinParallelCpgPass` ?

Comment (pandurangpatil) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/BasePassForAstProcessing.scala]: made the changes

Comment (DavidBakerEffendi) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/BasePassForAstProcessing.scala]: @pandurangpatil One would need to follow a similar chain of local builds as this PR: https://github.com/joernio/joern/pull/4693

**Review by DavidBakerEffendi — APPROVED** [review]

File: joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/BasePassForAstProcessing.scala (hunk 4 above)
Comment (mpollmeier) [reply joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/passes/BasePassForAstProcessing.scala]: Just a note: no need to adapt your changes here, I'm taking care to reapply them while rebasing. 

To make it simpler for you, you can just pretend that there's not flatgraph branch. But either way, pre-applying those changes didn't hurt either :)

**Pull request merged**